autogpts/autogpt/autogpt/agents/agent.py

```diff
@@ -64,157 +73,181 @@ class AgentConfiguration(BaseAgentConfiguration):
+    ) -> ThoughtProcessOutput:
+        ...
 
+
+class Agent(BaseAgent, Configurable[AgentSettings]):
     default_settings: AgentSettings = AgentSettings(
         name="Agent",
-        description=__doc__,
+        description=__doc__ if __doc__ else "",
     )
 
-    prompt_strategy: OneShotAgentPromptStrategy
-
     def __init__(
         self,
         settings: AgentSettings,
         llm_provider: ChatModelProvider,
-        command_registry: CommandRegistry,
         file_storage: FileStorage,
         legacy_config: Config,
     ):
-        prompt_strategy = OneShotAgentPromptStrategy(
-            configuration=settings.prompt_config,
-            logger=logger,
+        super().__init__(settings, llm_provider)
+
+        # Components
+        self.system = SystemComponent(legacy_config, settings.ai_profile)
+        self.history = EventHistoryComponent(
+            settings.history,
+            self.send_token_limit,
+            lambda x: self.llm_provider.count_tokens(x, self.llm.name),
+            legacy_config,
+            llm_provider,
+        )
+        self.user_interaction = UserInteractionComponent(legacy_config)
+        self.file_manager = FileManagerComponent(settings, file_storage)
+        self.code_executor = CodeExecutorComponent(
+            self.file_manager.workspace,
+            settings,
+            legacy_config,
         )
-        super().__init__(
-            settings=settings,
-            llm_provider=llm_provider,
-            prompt_strategy=prompt_strategy,
-            command_registry=command_registry,
-            file_storage=file_storage,
-            legacy_config=legacy_config,
+        self.git_ops = GitOperationsComponent(legacy_config)
+        self.image_gen = ImageGeneratorComponent(
+            self.file_manager.workspace, legacy_config
         )
+        self.web_search = WebSearchComponent(legacy_config)
+        self.web_selenium = WebSeleniumComponent(legacy_config, llm_provider, self.llm)
+        self.context = ContextComponent(self.file_manager.workspace)
+        self.watchdog = WatchdogComponent(settings.config, settings.history)
+        self.prompt_strategy = OneShotComponent(
+            legacy_config, llm_provider, self.send_token_limit, self.llm
+        )
+
+        # Override component ordering
+        self.components = [
+            self.system,
+            self.user_interaction,
+            self.file_manager,
+            self.code_executor,
+            self.history,
+            self.git_ops,
+            self.image_gen,
+            self.web_search,
+            self.web_selenium,
+            self.context,
+            self.watchdog,
+            self.prompt_strategy,
+        ]
 
         self.created_at = datetime.now().strftime("%Y%m%d_%H%M%S")
         """Timestamp the agent was created; only used for structured debug logging."""
 
         self.log_cycle_handler = LogCycleHandler()
         """LogCycleHandler for structured debug logging."""
 
```

Comment:
`build_prompt` and `execute` are still the two fundamental agent functions. I want to merge them into just one pipeline, and move all interaction (agent input and output) to the components. That would make it easier to use the agent outside, there will be no code necessary to externally handle passing data from `propose_action` to `execute`.